File: leetcode/45/45_jump_game_2.go
package main

import "fmt"

func jump(nums []int) int {
	l := len(nums)
	step := 0
	maxIdx := 0
	for i:=0;i<l-1;i++ {
		curMaxIdx := maxIdx
		for j:=i;j<=maxIdx;j++ {
			v := j + nums[j]
			if curMaxIdx < v {
				curMaxIdx = v
			}
		}
		step++
		i = maxIdx
		maxIdx = curMaxIdx
		if maxIdx >= l-1 {
			return step
		}
	}
	return step
}

func main() {
	// fmt.Println(jump([]int{7,0,9,6,9,6,1,7,9,0,1,2,9,0,3}))
	fmt.Println(jump([]int{1,2,3})) // 2
}

File: leetcode/45/45_jump_game_2.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    int jump(vector<int> &nums)
    {
        int l = nums.size();
        if (l <= 1)
        {
            return 0;
        }
        int maxIdx = 0;
        int step = 0;
        for (int i = 0; i < l - 1; i++)
        {
            int curMaxIdx = maxIdx;
            for (int j = i; j <= maxIdx; j++)
            {
                int v = j + nums[j];
                if (curMaxIdx < v)
                {
                    curMaxIdx = v;
                }
            }
            i = maxIdx;
            step++;
            maxIdx = curMaxIdx;
            if (maxIdx >= l - 1)
            {
                return step;
            }
        }
        return step;
    }
};

int main()
{
    Solution s;
    vector<int> v{2, 3, 1, 1, 4};
    // vector<int> v{1, 2, 3};
    cout << s.jump(v) << endl;
}